Source/CsDebugScript.Engine/Engine/Debuggers/DbgEngDll.cs

```diff
@@ -1243,6 +1260,312 @@ public DebugEventInfo GetLastEventInfo()
+
+                client = originalClient.CreateClient();
+
+                lock (eventCallbacksReady)
+                {
+                    debuggerStateLoop = 
+                        new System.Threading.Thread(() => DebuggerStateLoop(client, DebugStatusGo, DebugStatusBreak)) { IsBackground = true };
+                    debuggerStateLoop.SetApartmentState(System.Threading.ApartmentState.MTA);
+                    debuggerStateLoop.Start();
+
+                    // Wait for loop thread to become ready.
+                    //
+                    System.Threading.Monitor.Wait(eventCallbacksReady);
+                }
+            }
+
+            /// <summary>
+            /// Loop responsible to wait for debug events.
+            /// Needs to be run in separate thread.
+            /// </summary>
+            private static void DebuggerStateLoop(
+                IDebugClient client,
+                System.Threading.AutoResetEvent debugStatusGo,
+                System.Threading.AutoResetEvent debugStatusBreak)
+            {
+                bool hasClientExited = false;
+                var loopClient = client.CreateClient();
+                var eventCallbacks = new DebugCallbacks(loopClient, debugStatusGo);
+
+                lock (eventCallbacksReady)
+                {
+                    System.Threading.Monitor.Pulse(eventCallbacksReady);
+                }
+
+                // Default is to start in break mode, wait for release.
+                //
+                debugStatusGo.WaitOne();
+
+                while (!hasClientExited)
+                {
+                    ((IDebugControl7)loopClient).WaitForEvent(0, UInt32.MaxValue);
+                    uint executionStatus = ((IDebugControl7)loopClient).GetExecutionStatus();
+
+                    while (executionStatus == (uint)Defines.DebugStatusBreak)
+                    {
+                        debugStatusBreak.Set();
+                        debugStatusGo.WaitOne();
+
+                        executionStatus = ((IDebugControl7)loopClient).GetExecutionStatus();
+                    }
+
+                    hasClientExited = executionStatus == (uint)Defines.DebugStatusNoDebuggee;
+                }
+            }
+
+            /// <summary>
+            /// Waits for debugger loop thread to finish.
+            /// </summary>
+            public void WaitForDebuggerLoopToExit()
+            {
+                debuggerStateLoop.Join();
+            }
+        }
+
+        /// <summary>
+        /// Debug callbacks called during WaitForEvent callback.
+        /// This class in future can be extended to support callbacks provided
+        /// on certain actions (e.g. breakpoint hit, thread create, module load etc.)
+        /// </summary>
+        class DebugCallbacks : IDebugEventCallbacks
+        {
+            /// <summary>
+            /// IDebugClient.
+            /// </summary>
+            [ThreadStatic]
+            private IDebugClient client;
```

Comment:
Switched to just a reference without create new client.

---
In reply to: [69631897](https://github.com/southpolenator/WinDbgCs/pull/8#discussion_r69631897) [](ancestors = 69631897)